### ISSUE TITLE
parse error on input ‘-- * 'LHsDecl'

### ISSUE THREAD

**Issue by CantoroMC** [ndmitchell/hlint#1224]
I got this error while compiling

[237 of 264] Compiling GHC.Parser.PostProcess.Haddock ( compiler/GHC/Parser/PostProcess/Haddock.hs, dist/build/GHC/Parser/PostProcess/Haddock.o, dist/build/GHC/Parser/PostProcess/Haddock.dyn_o )

compiler/GHC/Parser/PostProcess/Haddock.hs:1469:3: error:
    parse error on input ‘-- * 'LHsDecl' produced by 'decl_cls' in Parser.y always have a 'BufSpan'’
     |
1469 |   -- * 'LHsDecl' produced by 'decl_cls' in Parser.y always have a 'BufSpan'
     |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
cabal: Failed to build ghc-lib-parser-9.0.1.20210207 (which is required by
exe:hlint from hlint-3.3).

**Comment by shayne-fletcher** [comment]
This is `ghc-lib-parser` build failure. This issue should be moved there. Would be helpful if you could provide the version of ghc you are building with please?

**Comment by shayne-fletcher** [comment]
It's not making a lot of sense. It has literally tripped up on a comment line. If your build compiler was 8.6.5 we'd have expected an earlier failure with a different message.

Interestingly, the comment has a `*` in it; this has been replaced on HEAD with `-`.
Before:
```
  -- * 'LHsDecl' produced by 'decl_cls' in Parser.y always have a 'BufSpan'
  -- * 'partitionBindsAndSigs' does not discard this 'BufSpan'
```
After:
```
  -- - 'LHsDecl' produced by 'decl_cls' in Parser.y always have a 'BufSpan'
  -- - 'partitionBindsAndSigs' does not discard this 'BufSpan'
```
One assumes there is therefore something dodgy about it then.

**Comment by CantoroMC** [comment]
GHC version 8.10.4

cabal-install version 3.2.0.0
compiled using version 3.2.1.0 of the Cabal library

I m running Arch Linux.
I've understood that the problem was related with ghc-lib-parser (i believed it was due to ghc-lib-parser-ex) but was not sure were to report the issue

**Comment by shayne-fletcher** [comment]
I'm really sorry you're having this problem! Unfortunately, I am struggling to get it reproduced. 

```
❯ cabal --version
cabal-install version 3.2.0.0
compiled using version 3.2.0.0 of the Cabal library 
❯ which ghc
/Users/shaynefletcher/ghc-8.10.4/bin/ghc
❯ cabal new-install hlint-3.3
Resolving dependencies...
Build profile: -w ghc-8.10.4 -O1
In order, the following will be built (use -v for more details):
 - base-compat-0.11.2 (lib) (requires build)
...
Starting     ghc-lib-parser-9.0.1.20210207 (lib)
...
Installing   ghc-lib-parser-9.0.1.20210207 (lib)
Completed    ghc-lib-parser-9.0.1.20210207 (lib)
...
Building     hlint-3.3 (exe:hlint)
Installing   hlint-3.3 (exe:hlint)
Completed    hlint-3.3 (exe:hlint)
```

**Comment by shayne-fletcher** [comment]
I have an idea for fixing this despite being unable to reproduce it. If someone could tell me if this script works?
```
#!/usr/bin/env bash

ghc_lib_parser=ghc-lib-parser-9.0.1.20210207
curl -o ${ghc_lib_parser}.tar.gz https://hackage.haskell.org/package/${ghc_lib_parser}/${ghc_lib_parser}.tar.gz
gunzip ${ghc_lib_parser}.tar.gz && tar xvf ${ghc_lib_parser}.tar
cd ${ghc_lib_parser}
#I'm on macOS which wants -i to take an extension argument, hence the ''
sed -i '' -e 's/-- \*/-- -/g' compiler/GHC/Parser/PostProcess/Haddock.hs # Fix the pesky comments in this file
cabal new-build
```
If it does, I'll release a patched version of ghc-lib-parser and we should be on our way.

**Comment by shayne-fletcher** [comment]
> I have an idea for fixing this despite being unable to reproduce it. If someone could tell me if this script works?
> 
> ```shell
> #!/usr/bin/env bash
> 
> ghc_lib_parser=ghc-lib-parser-9.0.1.20210207
> curl -o ${ghc_lib_parser}.tar.gz https://hackage.haskell.org/package/${ghc_lib_parser}/${ghc_lib_parser}.tar.gz
> gunzip ${ghc_lib_parser}.tar.gz && tar xvf ${ghc_lib_parser}.tar
> cd ${ghc_lib_parser}
> #I'm on macOS which wants -i to take an extension argument, hence the ''
> sed -i '' -e 's/-- \*/-- -/g' compiler/GHC/Parser/PostProcess/Haddock.hs # Fix the pesky comments in this file
> cabal new-build
> ```
> 
> If it does, I'll release a patched version of ghc-lib-parser and we should be on our way.

@Kleidukos tested this for us and found it worked. I'm preparing a new ghc-lib-parser release that incorporates the patch.

**Comment by shayne-fletcher** [comment]
ghc-lib-parser-9.0.1.20210324 is now on hackage

**Comment by shayne-fletcher** [comment]
@CantoroMC `cabal update&&cabal install` should solve the problem. do please let us know how it works out!

**Comment by CantoroMC** [comment]
works pretty pretty good

**Comment by shayne-fletcher** [comment]
nice 🎉 !

**Comment by ndmitchell** [comment]
Thanks @shayne-fletcher for fixing the bugs! I assume this issue can now be closed, but reopen if not.
